api/src/socket.rs

```diff
@@ -0,0 +1,241 @@
+}
+
+/// Copies a socket address from user space into a temporary kernel storage.
+///
+/// This function reads `addrlen` bytes from the user-space pointer `addr` and
+/// copies them into a `MaybeUninit<sockaddr>` in kernel memory.
+///
+#[inline]
+fn copy_sockaddr_from_user(
+    addr: UserConstPtr<sockaddr>,
+    addrlen: socklen_t,
+) -> LinuxResult<MaybeUninit<sockaddr>> {
+    let mut storage = MaybeUninit::<sockaddr>::uninit();
+    let sock_addr = addr.get_as_ref()?;
+    unsafe {
+        core::ptr::copy_nonoverlapping(
+            sock_addr as *const sockaddr as *const u8,
+            storage.as_mut_ptr() as *mut u8,
+            addrlen as usize,
+        )
+    };
+    Ok(storage)
+}
+
+impl SocketAddrExt for SocketAddr {
+    /// Reads a [`SocketAddr`] from user space.
+    ///
+    /// This implementation first performs basic length validation. Then, it copies
+    /// the raw [`sockaddr`] data from user space into a temporary kernel buffer.
+    /// Based on the address family ([`AF_INET`] or [`AF_INET6`]) extracted from the
+    /// copied data, it delegates the actual parsing to [`SocketAddrV4::read_from_user`]
+    /// or [`SocketAddrV6::read_from_user`].
+    fn read_from_user(addr: UserConstPtr<sockaddr>, addrlen: socklen_t) -> LinuxResult<Self> {
+        if size_of::<__kernel_sa_family_t>() > addrlen as usize
+            || addrlen as usize > size_of::<sockaddr>()
+        {
+            return Err(LinuxError::EINVAL);
+        }
+
+        let storage = copy_sockaddr_from_user(addr, addrlen)?;
+        let family = unsafe {
+            storage
+                .assume_init_ref()
+                .__storage
+                .__bindgen_anon_1
+                .__bindgen_anon_1
+                .ss_family as u32
+        };
+
+        match family {
+            AF_INET => SocketAddrV4::read_from_user(addr, addrlen).map(SocketAddr::V4),
+            AF_INET6 => SocketAddrV6::read_from_user(addr, addrlen).map(SocketAddr::V6),
+            _ => Err(LinuxError::EAFNOSUPPORT),
+        }
+    }
+
+    /// Writes the [`SocketAddr`] to user space.
+    ///
+    /// This implementation checks for a null user-space pointer. Then, it delegates
+    /// the actual writing to the specific [`SocketAddrV4`] or [`SocketAddrV6`]
+    /// `write_to_user` implementation based on the variant of `self`.
+    fn write_to_user(&self, addr: UserPtr<sockaddr>) -> LinuxResult<socklen_t> {
+        if addr.is_null() {
+            return Err(LinuxError::EINVAL);
+        }
+
+        match self {
+            SocketAddr::V4(v4) => v4.write_to_user(addr),
+            SocketAddr::V6(v6) => v6.write_to_user(addr),
+        }
+    }
+
+    /// Gets the address family of the [`SocketAddr`].
+    ///
+    /// Returns `AF_INET` for IPv4 addresses or `AF_INET6` for IPv6 addresses.
+    fn family(&self) -> u16 {
+        match self {
+            SocketAddr::V4(_) => AF_INET as u16,
+            SocketAddr::V6(_) => AF_INET6 as u16,
```

Comment:
Maybe it would be better to forward the implementation to the underlying type to reduce code duplication.